volatility/framework/plugins/windows/netlist.py

```diff
@@ -0,0 +1,544 @@
+#
+
+import logging
+import datetime
+from typing import Iterable, List, Optional, Callable
+
+from volatility.framework import constants, exceptions, interfaces, renderers, symbols, layers
+from volatility.framework.configuration import requirements
+from volatility.framework.renderers import format_hints
+from volatility.framework.symbols import intermed
+from volatility.framework.symbols.windows.extensions import network
+from volatility.framework.symbols.windows.pdbutil import PDBUtility
+from volatility.plugins import timeliner
+from volatility.plugins.windows import info, poolscanner, netscan, modules
+
+vollog = logging.getLogger(__name__)
+
+
+class NetList(interfaces.plugins.PluginInterface, timeliner.TimeLinerInterface):
+    """Scans for network objects present in a particular windows memory image."""
+
+    _required_framework_version = (2, 0, 0)
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls):
+        return [
+            requirements.TranslationLayerRequirement(name = 'primary',
+                                                     description = 'Memory layer for the kernel',
+                                                     architectures = ["Intel32", "Intel64"]),
+            requirements.SymbolTableRequirement(name = "nt_symbols", description = "Windows kernel symbols"),
+            requirements.VersionRequirement(name = 'netscan', component = netscan.NetScan, version = (1, 0, 0)),
+            requirements.BooleanRequirement(
+                name = 'include-corrupt',
+                description =
+                "Radically eases result validation. This will show partially overwritten data. WARNING: the results are likely to include garbage and/or corrupt data. Be cautious!",
+                default = False,
+                optional = True),
+        ]
+
+    @classmethod
+    def _decode_pointer(self, value):
+        """Copied from `windows.handles`.
+
+        Windows encodes pointers to objects and decodes them on the fly
+        before using them.
+
+        This function mimics the decoding routine so we can generate the
+        proper pointer values as well.
+        """
+
+        value = value & 0xFFFFFFFFFFFFFFFC
+
+        return value
+
+    @classmethod
+    def read_pointer(cls,
+                     context: interfaces.context.ContextInterface,
+                     layer_name: str,
+                     offset: int,
+                     length: int) -> int:
+        """Reads a pointer at a given offset and returns the address it points to.
+
+        Args:
+            context: The context to retrieve required elements (layers, symbol tables) from
+            layer_name: The name of the layer on which to operate
+            offset: Offset of pointer
+            length: Pointer length
+
+        Returns:
+            The value the pointer points to.
+        """
+
+        return int.from_bytes(context.layers[layer_name].read(offset, length), "little")
```

Comment:
Yeah, makes sense. There are some things in the current state which are not perfect, to say the least ;)

Which way would be best to overlay a pointer? Like this?

```python
part_table = context.object(net_symbol_table + constants.BANG + "pointer", 
                            layer_name = layer_name, 
                            offset = tcpip_module_offset + part_table_symbol)
```

Reply:
Yeah, although that defaults to being a pointer to void.  It might be worth defining a structure in the JSON which is then `pointer to actual structure` to make it easier to get around?  Otherwise I think you can do something like provide a `subtype`  parameter (which I think has to be a template, so use `get_type` to get it).

Reply:
But wouldn't that require a pointer subtype for each actual structure (e.g. `pointer to _INET_PORT_POOL`, `pointer to _PARTITION` etc.)? I'm tending to say that there are about 5 occurrences in `netlist` right now and I think we can work around that for now with void pointers. Of course if there is any big advantage to using proper subtypes we can change that at any point (or if something changes later on or it gets more complex) but right now I'm not really feeling the effort, you know? :S

Reply:
Oh nevermind, I understood what you meant. However, I've got an error. When updating the partition count of the partition table with this it does not really get updated...
```python
# part_table is the actual partition table offset and consists out of a dynamic amount of _PARTITION objects
# ~ part_table = context.object(net_symbol_table + constants.BANG + "_PARTITION_TABLE", 
                            # ~ layer_name = layer_name, 
                            # ~ offset = part_table_addr)
part_table = context.object(net_symbol_table + constants.BANG + "pointer",
                            layer_name = layer_name,
                            subtype = context.symbol_space.get_type(net_symbol_table + constants.BANG + "_PARTITION_TABLE"),
                            offset = tcpip_module_offset + part_table_symbol)
part_count = int.from_bytes(context.layers[layer_name].read(tcpip_module_offset + part_count_symbol, 1), "little")
part_table.Partitions.count = part_count
# part_count == 4
# part_table.Partitions.count is still 1 (default value)?
```

Reply:
Hmmm, ok, that really should work.  Would you mind filing me an issue so I don't forget to investigate it please?  5:S

Reply:
addressed in #400